File: nodejs/robot.js
const angle = require("./math/angle");
const Point = require("./math/point");
const Snapshot = require("./snapshot");
const { clamp } = require("./math/utils");

const MAX_SPEED = 10;

class Robot {
    role;
    snapshot;
    leftVelocity = 0;
    rightVelocity = 0;
    teamMessages = [];

    constructor(role) {
        this.role = role;
    }

    get rotation() { return this.snapshot.robot.rotation; }
    get position() { return this.snapshot.robot.position; }

    loop(data) {
        this.snapshot = new Snapshot(data);
        this.run();
        let teamMessages = this.teamMessages.slice();
        this.teamMessages.length = 0;
        return {
            team: teamMessages,
            L: this.leftVelocity,
            R: this.rightVelocity,
        };
    }

    sendDataToTeam(data) {
        this.teamMessages.push(data);
    }

    lookAtAngle(a) {
        let vl, vr;
        let ra = this.rotation;
        let delta = Math.min(angle.dist(a, ra), angle.dist(a, angle.opposite(ra)));
        let threshold = angle.degrees(1);

        if (delta < threshold) {
            vl = 0;
            vr = 0;
        } else {
            let vel = clamp(delta / angle.degrees(30), 0, 1);
            let p = Point.fromAngle(angle.radians(a - ra));
            if (p.x < 0) {
                vl = vel * -1;
                vr = vel;
            } else {
                vl = vel;
                vr = vel * -1;
            }
            if (p.y > 0) {
                vl *= -1;
                vr *= -1;
            }
        }
        
        this.leftVelocity = vl * MAX_SPEED;
        this.rightVelocity = vr * MAX_SPEED;
    }

    lookAtPoint(point) {
        let rx = this.position.x;
        let ry = this.position.y;
        let px = point.x;
        let py = point.y;
        this.lookAtAngle(new Point(px - rx, py - ry).angle);
    }

    moveToPoint(point) {
        let vl, vr;
        let rx = this.position.x;
        let ry = this.position.y;
        let px = point.x;
        let py = point.y;
        let a = new Point(px - rx, py - ry).angle;
        let ra = this.rotation;
        let delta = Math.min(angle.dist(a, ra), angle.dist(a, angle.opposite(ra)));
        let decrease = (angle.r2d(delta) / 90) * 2;
        let p = Point.fromAngle(angle.radians(a - ra));
        if (p.x < 0) {
            vl = 1 - decrease;
            vr = 1;
        } else {
            vl = 1;
            vr = 1 - decrease;
        }
        if (p.y > 0) {
            vl *= -1;
            vr *= -1;
        }
        
        this.leftVelocity = vl * MAX_SPEED;
        this.rightVelocity = vr * MAX_SPEED;
    }

    moveToBall() {
        this.moveToPoint(this.snapshot.ball.position);
    }

    run() {
        if (this.snapshot.isBallDetected()) {
            this.sendDataToTeam(this.snapshot.ball);
        }
        this.role.applyOn(this, this.snapshot);
    }
}

module.exports = Robot;

File: python/robot.py
import math_utils.angle as Angle
from math_utils.point import Point
from math_utils.utils import clamp
from snapshot import Snapshot

from collections import deque

MAX_SPEED = 10

class Robot:    
    leftVelocity = 0
    rightVelocity = 0
    teamMessages = deque()

    def __init__(self, role):
        self.role = role

    def getPosition(self):
        return self.snapshot.robot.position

    def getRotation(self):
        return self.snapshot.robot.rotation

    def loop(self, data):
        self.snapshot = Snapshot(data)
        self.run()
        teamMessages = list(self.teamMessages)
        self.teamMessages.clear()
        return {
            "team": teamMessages,
            "L": self.leftVelocity,
            "R": self.rightVelocity
        }

    def sendDataToTeam(self, data):
        self.teamMessages.append(data)

    def lookAtAngle(self, a):
        ra = self.getRotation()
        delta = min(Angle.diff(a, ra), Angle.diff(a, Angle.opposite(ra)))
        threshold = Angle.degrees(1)

        if delta < threshold:
            vl = 0
            vr = 0
        else:
            vel = clamp(delta / Angle.degrees(30), 0, 1)
            p = Point.fromAngle(Angle.radians(a - ra))
            if p.x < 0:
                vl = vel * -1
                vr = vel
            else:
                vl = vel
                vr = vel * -1
            if p.y > 0:
                vl = vl * -1
                vr = vr * -1
        
        self.leftVelocity = vl * MAX_SPEED
        self.rightVelocity = vr * MAX_SPEED

    def lookAtPoint(self, point):
        position = self.getPosition()
        rx = position.x
        ry = position.y
        px = point.x
        py = point.y
        self.lookAtAngle(Point(px - rx, py - ry).getAngle())

    def moveToPoint(self, point):
        position = self.getPosition()
        rx = position.x
        ry = position.y
        px = point.x
        py = point.y
        a = Point(px - rx, py - ry).getAngle()
        ra = self.getRotation()
        delta = min(Angle.diff(a, ra), Angle.diff(a, Angle.opposite(ra)))
        decrease = (Angle.r2d(delta) / 90) * 2
        p = Point.fromAngle(Angle.radians(a - ra))
        if p.x < 0:
            vl = 1 - decrease
            vr = 1
        else:
            vl = 1
            vr = 1 - decrease
        if p.y > 0:
            vl = vl * -1
            vr = vr * -1
        
        self.leftVelocity = vl * MAX_SPEED
        self.rightVelocity = vr * MAX_SPEED

    def moveToBall(self):
        self.moveToPoint(self.snapshot.ball.position)

    def run(self):
        if self.snapshot.isBallDetected():
            pos = self.snapshot.ball.position
            self.sendDataToTeam([pos.x, pos.y])

        self.role.applyOn(self, self.snapshot)